### PR TITLE
Attendees can sign up for attractions without badge numbers

### PR DESCRIPTION
**Pull request by EliAndrewC**
As discussed on the code@magfest.org mailing list earlier this week, this allows people to sign up for attractions before the event starts by entering their name/email/zipcode.

We originally talked about making this configurable on a per-feature basis, but it looks like Rob already added the ability for admins to toggle whether or not their attraction is ready for signups, which I believe handles that use case.

I tried a few different ways to add this, and I eventually settled on adding a "I don't have my badge yet" button to the signup modal:
newfooterbutton

which opens a different modal form:
newform

I spent some time trying to dynamically swap out the contents of the existing modal, but it was difficult to make the styling work due to the different sizes of the respective forms.  This approach works and doesn't involve any serious copy/paste.

I don't own a smartphone, but I tested this out in Google Chrome using their "Device Toolbar" in the Developer Tools.  The "I don't have my badge yet" button is a little big due to being wordy (happy to take suggestions for a less wordy name) but other than that everything looks okay to me.

### CODE REVIEW

**Comment by EliAndrewC** [comment]
Good catch!  I totally forgot about ``Attendee.normalized_email``, but I've just made that change and re-tested locally to make sure it works.

**Review by RobRuana — APPROVED** [review]
I am into this! I have a few thoughts, but I approve this as-is if you don't have time to make any more changes.

1. The naming for "early signup" could be swapped for "alt signup", as someone at the event could still use this method (for example if they left their badge in their room)
2. If we wanted to enable this per-attraction, I'd suggest a boolean column on the attraction table, like `enable_early_signups`. And then in the UI we would hide the "I don't have my badge" button if we're pre-fest and `enable_early_signups` is `False`. We'd probably also want to add a model validation on the backend. (Enabling per-attraction is actually a non-trivial amout of work)
3. I added the ability to hide attraction signups from the public, so it's probably less important that we enable this on a per-attraction basis. Admins can keep their attractions hidden from the public until the event starts 😄
4. After the signup is successful, we should probably still show the done modal: `$doneModal.modal('show');`. Maybe pass a varible to it that hides the "sign up a friend" form?

That's all I've got. Again, if you're short on time, I'm happy to make additional changes after the merge!

**Comment by RobRuana** [comment]
Also, from your screenshots, I'm glad to know the donut emoji "🍩" renders on linux!

File: panels/site_sections/attractions.py
Comment (kitsuta) [inline panels/site_sections/attractions.py]: We already have this -- it's `session.lookup_attendee` at https://github.com/magfest/ubersystem/blob/master/uber/models/__init__.py#L705. That version also handles duplicate attendees, which as we found can be pretty important. Looks like that version could stand to use normalized_email though.

File: panels/templates/attractions/events.html
Comment (kitsuta) [inline panels/templates/attractions/events.html:148]: This is cool, but from a quick Google search it looks like it only works in the latest Firefox. That's a bit too bleeding-edge -- unless I'm chasing the wrong lead and this isn't a destructuring assignment.

**Review by kitsuta — CHANGES_REQUESTED** [review]
This is super cool! I had a question about one of the JavaScript syntaxes you used, and we'd be better off if we used the existing attendee_lookup function, but I like this!

IMO, I think having attractions be private/public is good for now, and we don't need a config option for at-event/pre-event. In the future we can have a way for admins to schedule an attraction opening, which I think would completely cover the at-event-only use-case (and more!).

File: panels/templates/attractions/events.html
Comment (EliAndrewC) [reply panels/templates/attractions/events.html:148]: Ah, I see why you'd think that, but that's not what's happening here.

This is actually just passing an object literal as a parameter to this argument.  For example, instead of saying

```javascript
var d = {foo: 'bar'};
someFunc('baz', d, 'baf');
```

We can just say

```javascript
someFunc('baz', {foo: 'bar'}, 'baf'});
```

I can see why it looked like I was doing destructuring assignment, since what I did here looks a lot like the "Setting a function parameter's default value" section at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

However, in this case I'm only calling the function.  If I did something like this while *declaring* a function then you're right that it would only work in bleeding edge browsers and we should (sadly) avoid that kind of thing.

Comment (kitsuta) [reply panels/templates/attractions/events.html:148]: Ahh, yes that makes sense. It turns out googling "Javascript curly braces" isn't guaranteed to give context-aware results. Cool! Carry on!

File: panels/site_sections/attractions.py
Comment (EliAndrewC) [reply panels/site_sections/attractions.py]: Good catch, I've just made and tested this change.  You're right that version should use ``normalized_email``, so I've made that change at https://github.com/magfest/ubersystem/pull/3059

File: panels/site_sections/attractions.py
Comment (EliAndrewC) [inline panels/site_sections/attractions.py:26]: As per @kitsuta 's suggestion I've updated this function to use ``session.lookup_attendee``.

File: panels/templates/attractions/attractions_common.html
Comment (EliAndrewC) [inline panels/templates/attractions/attractions_common.html:217]: I implemented @RobRuana 's suggestion about showing the ``$doneModal`` instead of just a toast and hiding the badge form if we got there from the "I don't know my badge number" form.  I implemented that by adding an ``addClass`` / ``removeClass`` which sets or removes the ``hidden-modal-form`` class appropriately.

File: panels/templates/attractions/events.html
Comment (EliAndrewC) [inline panels/templates/attractions/events.html:78]: I liked @RobRuana 's suggestion of "alt" signup rather than "early" signup, so I renamed this to do that.

**Comment by EliAndrewC** [comment]
I've made all of the suggested changes except for the optional extra config.  In the interest of time I'm gonna merge this in and deploy it tonight since the tabletop folks have been asking for it.  If we find a need for that extra config between now and the event, let me know and I'll jump on it ASAP.

**Pull request merged**
